google/cloud/bigtable/data/_metrics/data_model.py

```diff
@@ -0,0 +1,514 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import annotations
+
+from typing import Callable, Any, Generator, Tuple, cast, TYPE_CHECKING
+
+import datetime
+import time
+import os
+import re
+import logging
+
+from enum import Enum
+from functools import lru_cache
+from dataclasses import dataclass
+from dataclasses import field
+from grpc import StatusCode
+
+import google.cloud.bigtable.data.exceptions as bt_exceptions
+
+if TYPE_CHECKING:
+    from google.cloud.bigtable.data._metrics.handlers._base import MetricsHandler
+
+
+# by default, exceptions in the metrics system are logged,
+# but enabling this flag causes them to be raised instead
+ALLOW_METRIC_EXCEPTIONS = os.getenv("BIGTABLE_METRICS_EXCEPTIONS", False)
+LOGGER = (
+    logging.getLogger(__name__) if os.getenv("BIGTABLE_METRICS_LOGS", False) else None
+)
+
+# default values for zone and cluster data, if not captured
+DEFAULT_ZONE = "global"
+DEFAULT_CLUSTER_ID = "unspecified"
+
+# keys for parsing metadata blobs
+BIGTABLE_METADATA_KEY = "x-goog-ext-425905942-bin"
+SERVER_TIMING_METADATA_KEY = "server-timing"
+SERVER_TIMING_REGEX = re.compile(r"gfet4t7; dur=(\d+)")
+
+INVALID_STATE_ERROR = "Invalid state for {}: {}"
+
+
+@dataclass(frozen=True)
+class TimeTuple:
+    """
+    Tuple that holds both the utc timestamp to record with the metrics, and a
+    monotonic timestamp for calculating durations. The monotonic timestamp is
+    preferred for calculations because it is resilient to clock changes, eg DST
+    """
+    utc: datetime.datetime = field(
+        default_factory=lambda: datetime.datetime.now(datetime.timezone.utc)
+    )
+    monotonic: float = field(default_factory=time.monotonic)
+
+
+class OperationType(Enum):
+    """Enum for the type of operation being performed."""
+
+    READ_ROWS = "ReadRows"
+    SAMPLE_ROW_KEYS = "SampleRowKeys"
+    BULK_MUTATE_ROWS = "MutateRows"
+    MUTATE_ROW = "MutateRow"
+    CHECK_AND_MUTATE = "CheckAndMutateRow"
+    READ_MODIFY_WRITE = "ReadModifyWriteRow"
+
+
+class OperationState(Enum):
+    """Enum for the state of the active operation."""
+
+    CREATED = 0
+    ACTIVE_ATTEMPT = 1
+    BETWEEN_ATTEMPTS = 2
+    COMPLETED = 3
+
+
+@dataclass(frozen=True)
+class CompletedAttemptMetric:
+    """
+    An immutable dataclass representing the data associated with a
+    completed rpc attempt.
+    """
+
+    start_time: datetime.datetime
```

Comment:
Should we include OperationType, cluster id, zone id, is streaming? Or those will be tagged on the metric later?

Reply:
These are the fields that are tracked for each individual attempt. Aren't those fields all tracked at the operation level?

Reply:
Yes, these are operation level labels and should be the same across multiple attempts. But Attempts are also labeled with these fields, so want to make sure they're added to the attributes later :)

Reply:
Right now, `on_attempt_complete(attempt, operation)` is passed a `CompletedAttemptMetric` and an `ActiveOperationMetric`, and between the two of them, the attempt callback will have access to all the fields we need. I'm hesitant to duplicate data between them, because it means more tests and more room for errors (not to mention more slow Python lines that need to be executed)

My thinking is that some data lives at the individual attempt level, and some is shared at the operation level, so I tried to structure the data models in that way. But let me know if I'm misunderstanding something